### PR TITLE
Open MongoDb Data API, Add ObjectID Scalar & Filtering

### LINKED ISSUE #2534: Filtering by ObjectId type with HotChocolate
I am trying to create a Graphql Schema and schema is crating but I cannot see a filtering option for Id (which is MongoDB ObjectId type). Could you tell me what I am missing? I would like to filter by Id as well but there is no like that option on the created schema. You can see my test implementation below for that.
Note: The HotChocolate Version is 10.5.3

```
public class Startup
    {
        public IConfiguration Configuration { get; }

        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            TypeConversion.Default.Register<string, ObjectId>(from => ObjectId.Parse(from));
            TypeConversion.Default.Register<ObjectId, string>(from => from.ToString());            
            services.AddSingleton<IAuthorService, InMemoryAuthorService>();

            services.AddGraphQL(s => SchemaBuilder.New()
                .AddServices(s)
                .AddType<AuthorType>()
                .AddQueryType<Query>()
                .BindClrType<ObjectId, IdType>()
                .Create());
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UsePlayground(new PlaygroundOptions
                {
                    QueryPath = "/api",
                    Path = "/playground"
                });
            }
            app.UseRouting();
            app.UseGraphQL("/api");
        }
    }
 public class AuthorType : ObjectType<Author>
    {
        protected override void Configure(IObjectTypeDescriptor<Author> descriptor)
        {
            descriptor.Field(t => t.Id)
                     .Type<IdType>()
                     .Resolver(c => c.Parent<Author>().Id);
            descriptor.Field(a => a.AuthorId).Type<StringType>();
            descriptor.Field(a => a.Name).Type<StringType>();
            descriptor.Field(a => a.Surname).Ignore();
        }
    }
    public class Author
    {
        public ObjectId Id { get; set; }
        public int AuthorId { get; set; }
        public string Name { get; set; }
        public string Surname { get; set; }
    }

    public class Query
    {
        private readonly IAuthorService _authorService;            

        public Query(IAuthorService authorService)
        {
            _authorService = authorService;               
        }
        [UsePaging]
        [UseFiltering]
        public IQueryable<Author> Authors() => _authorService.GetAll();    
      
    }


    public interface IAuthorService
    {
        IQueryable<Author> GetAll();
    }

    public class InMemoryAuthorService : IAuthorService
    {
        private IList<Author> _authors;

        public InMemoryAuthorService()
        {
            _authors = new List<Author>()
            {
                new Author() {Id = ObjectId.Parse("5e4442c28ae77cb23386b911"), AuthorId = 1, Name = "Name 1", Surname = "Surname 1"},
                new Author() {Id = ObjectId.Parse("5e4442c28ae77cb23386b912"), AuthorId = 2, Name = "Name 2", Surname = "Surname 2"},
                new Author() {Id = ObjectId.Parse("5e4442c28ae77cb23386b913"),AuthorId = 3, Name = "Name 4", Surname = "Surname 3"}
            };
        }

        public IQueryable<Author> GetAll()
        {
            return _authors.AsQueryable();
        }
    }
```

image

### PR DESCRIPTION
**Pull request by PascalSenn**
- Opens MongoDB Data API for extensions
- Adds ObjectId Scalar
- Adds ObjectId Filters

Closes #2534

### CODE REVIEW

**Comment by PascalSenn** [comment]
@tobias-tengler 
I run into issues with the analyzer


<details><summary>Error</summary>
<p>

```
Repeating warnings and errors:
Check failed: Could not complete checking build configurations within 500 milliseconds.
Check failed: GitVersion does not support fetching SSH endpoints. Enable NoFetch to skip fetching.
Warnings were encountered while loading the workspace. Set the verbosity option to the 'diagnostic' level to log warnings.
Unhandled exception: System.NotSupportedException: Adding additional documents is not supported.
   at Microsoft.CodeAnalysis.Workspace.CheckAllowedProjectChanges(ProjectChanges projectChanges)
   at Microsoft.CodeAnalysis.Workspace.CheckAllowedSolutionChanges(SolutionChanges solutionChanges)
   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)
   at Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)
   at Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.TryApplyChanges(Solution newSolution)
   at Microsoft.CodeAnalysis.Tools.CodeFormatter.FormatWorkspaceAsync(FormatOptions formatOptions, ILogger logger, CancellationToken cancellationToken, String binaryLogPath) in /_/src/CodeFormatter.cs:line 99
   at Microsoft.CodeAnalysis.Tools.Program.Run(String workspace, Boolean noRestore, Boolean folder, Boolean fixWhitespace, String fixStyle, String fixAnalyzers, String[] diagnostics, String verbosity, Boolean check, String[] include, String[] exclude, String report, Boolean includeGenerated, String binarylog, IConsole console) in /_/src/Program.cs:line 185
   at System.CommandLine.Invocation.CommandHandler.GetExitCodeAsync(Object value, InvocationContext context)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass27_0.<<UseVersionOption>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__24_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseDebugDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
ProcessException: Process 'dotnet' exited with code 1.
   > /usr/local/share/dotnet/dotnet format "/Users/pascalsenn/kot/hotchocolate/src/Build.CheckApi.sln" --fix-analyzers warn --diagnostics RS0016
   @ /Users/pascalsenn/kot/hotchocolate
Error output:
   Warnings were encountered while loading the workspace. Set the verbosity option to the 'diagnostic' level to log warnings.
   Unhandled exception: System.NotSupportedException: Adding additional documents is not supported.
      at Microsoft.CodeAnalysis.Workspace.CheckAllowedProjectChanges(ProjectChanges projectChanges)
      at Microsoft.CodeAnalysis.Workspace.CheckAllowedSolutionChanges(SolutionChanges solutionChanges)
      at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)
      at Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)
      at Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.TryApplyChanges(Solution newSolution)
      at Microsoft.CodeAnalysis.Tools.CodeFormatter.FormatWorkspaceAsync(FormatOptions formatOptions, ILogger logger, CancellationToken cancellationToken, String binaryLogPath) in /_/src/CodeFormatter.cs:line 99
      at Microsoft.CodeAnalysis.Tools.Program.Run(String workspace, Boolean noRestore, Boolean folder, Boolean fixWhitespace, String fixStyle, String fixAnalyzers, String[] diagnostics, String verbosity, Boolean check, String[] include, String[] exclude, String report, Boolean includeGenerated, String binarylog, IConsole console) in /_/src/Program.cs:line 185
      at System.CommandLine.Invocation.CommandHandler.GetExitCodeAsync(Object value, InvocationContext context)
      at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
      at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
   --- End of stack trace from previous location ---
      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseParseErrorReporting>b__0>d.MoveNext()
   --- End of stack trace from previous location ---
      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
   --- End of stack trace from previous location ---
      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass27_0.<<UseVersionOption>b__1>d.MoveNext()
   --- End of stack trace from previous location ---
      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseTypoCorrections>b__0>d.MoveNext()
   --- End of stack trace from previous location ---
      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__24_0>d.MoveNext()
   --- End of stack trace from previous location ---
      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseParseDirective>b__0>d.MoveNext()
   --- End of stack trace from previous location ---
      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseDebugDirective>b__0>d.MoveNext()
   --- End of stack trace from previous location ---
      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
   --- End of stack trace from previous location ---
      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

</p>
</details>

Do i have to do more than just `nuke AddUnshippedAPI`?

**Comment by tobias-tengler** [comment]
Yes, you added a new project MongoDB.Types, but didn't copy over the PubliAPI.empty.txt file. So PublicAPI.Shipped.txt and PublicAPI.Unshipped.txt were missing.
I'll add them + the unshipped APIs.

**Comment by PascalSenn** [comment]
@tobias-tengler  Thanks tobias :)

**Comment by PascalSenn** [comment]
Drop Scalar

File: src/HotChocolate/MongoDb/src/Types/MongoDbScalarNames.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+namespace HotChocolate.Types.MongoDb
+{
+    public static class MongoDbScalarNames
```
Comment (michaelstaib) [inline src/HotChocolate/MongoDb/src/Types/MongoDbScalarNames.cs:3]: XML Docs

**Review by michaelstaib — APPROVED** [review]

**Pull request merged**